src/main/java/net/neoforged/neoforge/attachment/DataAttachmentOps.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.attachment;
+
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.DynamicOps;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+import net.minecraft.core.HolderLookup;
+import net.minecraft.resources.DelegatingOps;
+import net.minecraft.resources.RegistryOps;
+import net.minecraft.util.ExtraCodecs;
+
+public class DataAttachmentOps<T, TOwner> extends DelegatingOps<T> {
+    private final HolderLookup.Provider holderLookup;
+    private final TOwner parent;
+
+    protected DataAttachmentOps(RegistryOps<T> regOps, HolderLookup.Provider holderLookup, TOwner parent) {
+        super(regOps);
+        this.holderLookup = holderLookup;
+        this.parent = parent;
+    }
+
+    public static <T, TOwner extends IAttachmentHolder> DataAttachmentOps<T, TOwner> create(HolderLookup.Provider holderLookup, DynamicOps<T> targetOps, TOwner parent) {
+        return new DataAttachmentOps<>(holderLookup.createSerializationContext(targetOps), holderLookup, parent);
+    }
+
+    public HolderLookup.Provider registryLookup() {
+        return holderLookup;
+    }
+
+    public static <O, TOwner extends IAttachmentHolder> RecordCodecBuilder<O, TOwner> holder(Class<TOwner> holder) {
+        return ExtraCodecs.retrieveContext(ops -> {
+            if (ops instanceof DataAttachmentOps<?, ?> dops) {
+                if (holder.isInstance(dops.parent))
```

Comment:
Is this the only place that hackily-retrieved parent class is actually used? It seems to be best I can tell. I do seriously wonder if it's not better just to have the class passed in explicitly in the, like, one place it's actually grabbed from -- the constructor of `AttachmentHolder`.

Reply:
`DataAttachmentOps` is the approach any attachment user can take to capture the holder reference in their codecs. It's an automatic re-attach of the parent; see the tests.